File: web-ide/rsc-highlighter.js
(function () {

    const style = document.createElement("style");
    style.innerHTML = `
        .token.comment { color: #6a9955; font-style: italic; }
        .token.string { color: #ce9178; }
        .token.keyword { color: #569cd6; font-weight: bold; }
        .token.storage-type { color: #4ec9b0; }
        .token.storage-modifier { color: #c586c0; }
        .token.register { color: #9cdcfe; }
        .token.number { color: #b5cea8; }
        .token.operator { color: #d4d4d4; }
        .token.punctuation { color: #808080; }
        .token.directive { color: #c586c0; font-weight: bold; }
        .token.function-def { color: #dcdcaa; font-weight: bold; }
        .token.function-call { color: #dcdcaa; }
        .token.label-def { color: #4fc1ff; text-decoration: underline; font-weight: bold; }
        .token.label-ref { color: #4fc1ff; }
        .token.python-func { color: #dcdcaa; }
        .token.support-variable { color: #9cdcfe; font-style: italic; }

        .token.string .token.parameter { color: #4fc1ff; font-weight: bold; }
        .token.string .token.escape { color: #d7ba7d; }
    `;
    document.head.appendChild(style);


    window.initRscHighlighter = function(syntaxData) {
        if (typeof Prism === "undefined") return;

        let rsc = {
            // Static rules (not in syntax.json since they use specific Prism features like inside/greedy)
            comment: [
                { pattern: /\/\*[\s\S]*?\*\//, greedy: true },
                { pattern: /#.*/, greedy: true }
            ],
            string: [
                {
                    pattern: /"(?:\\.|[^"\\])*"/,
                    greedy: true,
                    inside: {
                        parameter: { pattern: /\{[^}]+\}/, alias: "variable" },
                        escape: /~/
                    }
                },
                { pattern: /'(?:\\.|[^'\\])*'/, greedy: true }
            ]
        };

        const ALIAS_MAP = {
            "directive": "important",
            "support_variable": "variable",
            "distance_helper": "keyword",
            "builtin": "function",
            "python_func": "function",
            "function_def": "function",
            "function_call": "function",
            "function_call_direct": "function",
            "register": "variable",
            "number_hex_array": "number",
            "number_hex": "number",
            "number_hex_byte": "number",
            "number_dec": "number"
        };

        let tempRules = {};

        for (let rule of syntaxData.rules) {
            let id = rule.id;
            let regexStr = rule.regex;
            let flags = rule.flags_js || "";
            let obj = { pattern: new RegExp(regexStr, flags) };
            if (rule.group === 2) obj.lookbehind = true;
            if (ALIAS_MAP[id]) obj.alias = ALIAS_MAP[id];
            
            tempRules[id] = obj;
        }

        // Map parsed rules to Prism token names
        rsc["directive"] = tempRules["directive"];
        rsc["support-variable"] = tempRules["support_variable"];
        rsc["storage-modifier"] = tempRules["storage_modifier"];
        rsc["storage-type"] = tempRules["storage_type"];
        rsc["keyword"] = tempRules["keyword"];
        rsc["distance-helper"] = tempRules["distance_helper"];
        rsc["builtin"] = tempRules["builtin"];
        rsc["python-func"] = tempRules["python_func"];
        rsc["function-def"] = tempRules["function_def"];
        rsc["function-call"] = [tempRules["function_call"], tempRules["function_call_direct"]];
        rsc["label-def"] = [tempRules["label_def_1"], tempRules["label_def_2"]];
        rsc["label-ref"] = [tempRules["label_ref_1"], tempRules["label_ref_2"]];
        rsc["register"] = tempRules["register"];
        rsc["constant"] = tempRules["constant"];
        rsc["number"] = [tempRules["number_hex"], tempRules["number_hex_array"], tempRules["number_hex_byte"], tempRules["number_dec"]];
        rsc["operator"] = tempRules["operator"];
        rsc["punctuation"] = tempRules["punctuation"];

        Prism.languages.rsc = rsc;
    };

})();

File: tui-ide/rsc_highlighter.py
"""
RSC / ASM syntax highlighter for the TUI IDE.
Mirrors the rules from .vscode/extensions/rsc-syntax/syntaxes/rsc.tmLanguage.json.

Usage:
    from rsc_highlighter import highlight_rsc_line
    spans = highlight_rsc_line(text)  # list of (start, end, rich.style.Style)
"""
import re
import json
import os
from functools import lru_cache
from rich.style import Style

# ----------------------------------------------------------------------
# 1. Scope/Style Mappings
# ----------------------------------------------------------------------
S_COMMENT    = Style(color="#6272a4", italic=True)   # Comments
S_STRING     = Style(color="#f1fa8c")                # Strings
S_KEYWORD    = Style(color="#ff79c6", bold=True)     # Control flow
S_KEYWORD_OP = Style(color="#ff79c6")                # in, and, or, not
S_STORAGE    = Style(color="#ff79c6", bold=True)     # var, reg, loop, repeat, str
S_NUMBER     = Style(color="#bd93f9")                # Numbers (dec, hex)
S_HEX_BYTE   = Style(color="#bd93f9", bold=True)     # Hex bytes (FF, 0A)
S_REGISTER   = Style(color="#50fa7b")                # Registers
S_FUNCTION   = Style(color="#50fa7b", bold=True)     # Function names
S_BUILTIN    = Style(color="#8be9fd")                # Built-in functions
S_LABEL      = Style(color="#f8f8f2", underline=True)# Labels definition
S_LABEL_REF  = Style(color="#ffb86c")                # Label references
S_DIRECTIVE  = Style(color="#ff79c6", bold=True)     # @build, @set, @section
S_SUPPORT    = Style(color="#8be9fd")                # emu.*, line.*, output.*
S_OPERATOR   = Style(color="#ff79c6")                # ==, !=, +, -
S_PUNCT      = Style(color="#f8f8f2")                # [, ], {, }, (, )
S_KEY        = Style(color="#ffb86c", bold=True)     # KEY_xxx
S_DEFAULT    = Style(color="#f8f8f2")                # Normal text

# ----------------------------------------------------------------------
# 2. Regex Patterns (Loaded from JSON)
# ----------------------------------------------------------------------
_PATTERNS: list[tuple[re.Pattern, int | None, Style]] = []

STYLE_MAP = {
    "directive": S_DIRECTIVE,
    "support_variable": S_SUPPORT,
    "storage_modifier": S_STORAGE,
    "storage_type": S_STORAGE,
    "keyword": S_KEYWORD,
    "distance_helper": S_KEYWORD,
    "builtin": S_BUILTIN,
    "python_func": S_FUNCTION,
    "function_def": S_FUNCTION,
    "function_call": S_FUNCTION,
    "function_call_direct": S_FUNCTION,
    "label_def_1": S_LABEL,
    "label_def_2": S_LABEL,
    "label_ref_1": S_LABEL_REF,
    "label_ref_2": S_LABEL_REF,
    "register": S_REGISTER,
    "constant": S_KEY,
    "number_hex_array": S_HEX_BYTE,
    "number_hex": S_NUMBER,
    "number_hex_byte": S_HEX_BYTE,
    "number_dec": S_NUMBER,
    "operator": S_OPERATOR,
    "punctuation": S_PUNCT
}

def load_syntax():
    json_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), "libcompiler", "syntax.json")
    try:
        with open(json_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        for rule in data.get("rules", []):
            rule_id = rule.get("id")
            regex_str = rule.get("regex")
            if not rule_id or not regex_str or rule_id not in STYLE_MAP:
                continue
            flags = 0
            if rule.get("flags_py") == "IGNORECASE":
                flags |= re.IGNORECASE
            # flags_js='m' could map to re.MULTILINE if needed, but we match line by line in Python anyway
            grp = rule.get("group")
            _PATTERNS.append((re.compile(regex_str, flags), grp, STYLE_MAP[rule_id]))
    except Exception as e:
        print(f"Failed to load syntax.json: {e}")

load_syntax()


@lru_cache(maxsize=10000)
def highlight_rsc_line(text: str, in_comment: bool = False) -> tuple[list[tuple[int, int, Style]], bool]:
    """
    Returns a list of (start, end, Style) spans for the given line, and the new in_comment state.
    Spans do not overlap; earlier patterns win (priority order).
    """
    n = len(text)
    
    claimed = bytearray(n)  

    results: list[tuple[int, int, Style]] = []

    
    pos = 0
    while pos < n:
        if in_comment:
            end_idx = text.find('*/', pos)
            if end_idx != -1:
                end_comment = end_idx + 2
                results.append((pos, end_comment, S_COMMENT))
                for i in range(pos, end_comment): claimed[i] = 1
                pos = end_comment
                in_comment = False
            else:
                results.append((pos, n, S_COMMENT))
                for i in range(pos, n): claimed[i] = 1
                in_comment = True
                break
        else:
            
            next_block = text.find('/*', pos)
            next_str1 = text.find('"', pos)
            next_str2 = text.find("'", pos)
            next_line_comment = text.find('#', pos)
            
            candidates = []
            if next_block != -1: candidates.append((next_block, 'block'))
            if next_str1 != -1: candidates.append((next_str1, 'str1'))
            if next_str2 != -1: candidates.append((next_str2, 'str2'))
            if next_line_comment != -1: candidates.append((next_line_comment, 'line'))
            
            if not candidates:
                break
                
            first_idx, kind = min(candidates, key=lambda x: x[0])
            
            if kind == 'block':
                in_comment = True
                pos = first_idx
            elif kind == 'str1':
                end_idx = first_idx + 1
                while end_idx < n:
                    if text[end_idx] == '"' and text[end_idx-1] != '\\':
                        break
                    end_idx += 1
                end_idx = min(end_idx + 1, n)
                results.append((first_idx, end_idx, S_STRING))
                for i in range(first_idx, end_idx): claimed[i] = 1
                pos = end_idx
            elif kind == 'str2':
                end_idx = first_idx + 1
                while end_idx < n:
                    if text[end_idx] == "'" and text[end_idx-1] != '\\':
                        break
                    end_idx += 1
                end_idx = min(end_idx + 1, n)
                results.append((first_idx, end_idx, S_STRING))
                for i in range(first_idx, end_idx): claimed[i] = 1
                pos = end_idx
            elif kind == 'line':
                results.append((first_idx, n, S_COMMENT))
                for i in range(first_idx, n): claimed[i] = 1
                break

    for pattern, grp, style in _PATTERNS:
        for m in pattern.finditer(text):
            if grp is not None:
                try:
                    start, end = m.span(grp)
                except IndexError:
                    start, end = m.span(0)
            else:
                start, end = m.span(0)

            if start >= n or end > n:
                continue
            
            if any(claimed[start:end]):
                continue
            
            for i in range(start, end):
                claimed[i] = 1
            results.append((start, end, style))

    results.sort(key=lambda x: x[0])
    return results, in_comment


def make_segments(line_text: str, in_comment: bool = False, default_style: Style = S_DEFAULT) -> tuple[list, bool]:
    """
    Convert a plain text line + highlight spans into a list of rich Segment objects.
    """
    from rich.segment import Segment

    spans, next_in_comment = highlight_rsc_line(line_text, in_comment)
    segments = []
    pos = 0
    for start, end, style in spans:
        if pos < start:
            segments.append(Segment(line_text[pos:start], default_style))
        segments.append(Segment(line_text[start:end], style))
        pos = end
    if pos < len(line_text):
        segments.append(Segment(line_text[pos:], default_style))
    return segments, next_in_comment
